### PR TITLE
[BPK-843] React Native support with BpkText component

### PR DESCRIPTION
**Pull request by shaundon**
+ [x] For any new components I've made sure that the style can be extended by the consumer using a `className` override.
+ [x] For any new files I've included the [Apache 2.0 License header](https://github.com/Skyscanner/backpack/blob/master/packages/bpk-tokens/formatters/license-header.js#L2-L16) at the top of the file.
+ [x] I've updated `changelog.md` for any changes that need to be highlighted in the changelog.
+ [x] If I've updated `npm-shrinkwrap.json` those changes are intentional.

This PR adds the basic React Native config, plus a single component (a port of `BpkText`) to Backpack.

In summary:
- Lerna now looks in `native/packages/*` for RN packages.
- Almost everything outside of `packages/` in the `native` folder is concerned with running the native code locally, using Storybook.
- All the RN code can be run from the root of the repo.

Caveat(s):
- Right now, running RN locally is a two step process. One step to run storybook in the background, then another to run RN on device. This means you need two console tabs open. We may wish to streamline this in the future.

### CODE REVIEW

File: native/android/app/src/main/java/com/native/MainActivity.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package com.skyscanner;
```
Comment (shaundon) [inline native/android/app/src/main/java/com/native/MainActivity.java]: Quick note - as the folder the app lives in is called `native`, this was originally `com.native`. However this wouldn't build because it's apparently an invalid package name, so I brute forced my way out of it by changing it to `com.skyscanner` everywhere it occurred.

File: changelog.md [hunk 2]
```diff
@@ -8,6 +8,9 @@
   - Added `BpkBackgroundImage` component which can be used to create a container with a background image
   - Added `withLazyLoading` HOC which allows `BpkImage` or `BpkBackgroundImage` to be loaded only when in view
   - Added `withLoadingBehavior` HOC which allows `BpkImage` or `BpkBackgroundImage` to be shown only once loaded, displaying a spinner until that point
+- react-native-bpk-component-text: 0.0.1-alpha.1
```
Comment (matthewdavidson) [inline changelog.md]: The `react-native-bpk-component-text` version should be left blank in the changelog (We'll publish it at 1.0.0 for first proper release anyways).

Comment (matthewdavidson) [reply changelog.md]: We'll drop the alpha comment too ;).

File: contributing.md [hunk 3]
```diff
@@ -222,3 +222,82 @@ When adding documentation for a new component, add the new dependency in `packag
 - Update the [changelog](/changelog.md) with the package versions and update the title with today’s date and a brief summary of what has changed.
 - Commit and push to master.
 - Update the #backpack Slack channel and the [BackpackNeedToKnow](mailto:backpackneedtoknow@skyscanner.net) mailing list if necessary.
+
+## React Native
```
Comment (matthewdavidson) [inline contributing.md]: This section should come before "Publishing packages" I reckon. Also we should add an item to the table of contents at the top of this file.

File: native/.flowconfig [hunk 4]
```diff
@@ -0,0 +1,45 @@
+[ignore]
```
Comment (matthewdavidson) [inline native/.flowconfig]: we don't use flow (yet...), so this file can be deleted ..?

File: native/.gitignore [hunk 5]
```diff
@@ -0,0 +1,53 @@
+# OSX
```
Comment (matthewdavidson) [inline native/.gitignore]: Kinda wondering if we need all of these ...hard to tell mind you 🤔 .

File: native/.watchmanconfig [hunk 6]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (matthewdavidson) [inline native/.watchmanconfig]: empty config? what happens if we delete it ?

File: native/android/app/src/main/java/com/native/MainActivity.java (hunk 1 above)
Comment (matthewdavidson) [reply native/android/app/src/main/java/com/native/MainActivity.java]: perhaps `com.backpack`?

File: native/package.json [hunk 7]
```diff
@@ -1,9 +1,41 @@
...
+  "scripts": {
+    "start": "node node_modules/react-native/local-cli/cli.js start",
+    "ios": "node node_modules/react-native/local-cli/cli.js run-ios",
+    "android": "node node_modules/react-native/local-cli/cli.js run-android",
+    "test": "jest",
+    "storybook": "storybook start -p 7007"
+  },
+  "dependencies": {
+    "react": "16.0.0-alpha.12",
+    "react-native": "0.47.2"
```
Comment (matthewdavidson) [inline native/package.json]: we need to make sure these are the same as what's used in the internal iOS and android apps.

Comment (matthewdavidson) [reply native/package.json]: Plus I reckon they can be `devDependencies`.

File: package.json [hunk 8]
```diff
@@ -31,7 +32,11 @@
     "docs:dist": "npm run clean:dist && NODE_ENV=production webpack && npm run sassdoc",
     "docs:dist:gzip": "npm run docs:dist && gulp",
     "start": "npm run build && npm run storybook",
-    "test": "npm run lint && npm run jest",
+    "start:native": "(cd ./native && npm run storybook)",
```
Comment (matthewdavidson) [inline package.json]: Perhaps just `"native"` instead of `"start:native"`?

File: native/android/app/src/main/java/com/native/MainActivity.java (hunk 1 above)
Comment (shaundon) [reply native/android/app/src/main/java/com/native/MainActivity.java]: Can do that, sure!

Comment (shaundon) [reply native/android/app/src/main/java/com/native/MainActivity.java]: Done

File: changelog.md (hunk 2 above)
Comment (shaundon) [reply changelog.md]: Done

File: contributing.md (hunk 3 above)
Comment (shaundon) [reply contributing.md]: Done - I put this at the bottom initially as I couldn't decide where it would best fit, and forgot to move it up before making the PR.

File: native/.flowconfig (hunk 4 above)
Comment (shaundon) [reply native/.flowconfig]: I'll try it and see what happens, should be fine.

File: native/.gitignore (hunk 5 above)
Comment (shaundon) [reply native/.gitignore]: These were all autogenerated, I'll remove the ones I know we won't need for sure.

File: package.json (hunk 8 above)
Comment (shaundon) [reply package.json]: I'm not sure 🤔  I went with this way to match the convention in the other scripts e.g. `postinstall / postinstall:native`, `test / test:native`, so it would be good to preserve that. It would be cleaner just to do `npm run native` though..

File: native/.watchmanconfig (hunk 6 above)
Comment (shaundon) [reply native/.watchmanconfig]: Let's find out!

Comment (shaundon) [reply native/.watchmanconfig]: It seems to be fine with it deleted!

File: native/.flowconfig (hunk 4 above)
Comment (shaundon) [reply native/.flowconfig]: Done

File: native/storybook/stories/index.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable import/no-extraneous-dependencies, import/no-unresolved, import/extensions */
```
Comment (matthewdavidson) [inline native/storybook/stories/index.js]: Can we move that actual stories to live inside component directories?

i.e. line 22 of this file could live in `native/packages/react-native-bpk-component-text/stories.js`

Comment (shaundon) [reply native/storybook/stories/index.js]: Yep, that makes sense, will do that.

**Comment by matthewdavidson** [comment]
@shaundon I worked through the instructions for android and everything worked as expected. This is looking really good 👍 .

File: native/.gitignore (hunk 5 above)
Comment (shaundon) [reply native/.gitignore]: I've removed a couple of things, but I think it might be best to keep the rest in there until we know for sure what we don't need - particularly the XCode stuff, as I can see a lot of these files mysteriously appearing if we open the project in XCode.

File: native/storybook/stories/index.js (hunk 9 above)
Comment (shaundon) [reply native/storybook/stories/index.js]: Done

File: package.json (hunk 8 above)
Comment (matthewdavidson) [reply package.json]: The convention is `task:subtask` i.e. by virtue of running `task` all `subtasks` are executed.

`native` isn't a subtask of `start`, they are mutually exclusive.

Comment (shaundon) [reply package.json]: Ahh ok, I didn't realise that was the convention.

File: native/.babelrc [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (georgegillams) [inline native/.babelrc]: Should these configs have our standard license header???

File: package.json (hunk 8 above)
Comment (shaundon) [reply package.json]: Done

File: native/.babelrc (hunk 10 above)
Comment (shaundon) [reply native/.babelrc]: The ones in the root of the folder don't have it, so I matched that here. Not to say that we shouldn't have them though.

Maybe we should decide amongst ourselves which files need to have it and then any changes to be made can be done in another PR to keep it separate?

Comment (matthewdavidson) [reply native/.babelrc]: 👍

File: native/package.json (hunk 7 above)
Comment (shaundon) [reply native/package.json]: Just looked at the iOS and Android apps, iOS matches this, Android is on `16.0.0-alpha.8`. I reckon therefore we can stick with this. The version of RN is a bit less clear cut due to the fork, but I understand that this is outside of the scope of this PR.

File: native/.babelrc (hunk 10 above)
Comment (matthewdavidson) [reply native/.babelrc]: json documents don't support comments ;)

Comment (shaundon) [reply native/.babelrc]: This is why NPM needs support for a `package.js` file!

File: readme.md [hunk 11]
```diff
@@ -66,6 +66,7 @@ npm install [package-name] --save-dev
 | [`bpk-stylesheets`](/packages/bpk-stylesheets) | [![npm version](https://badge.fury.io/js/bpk-stylesheets.svg)](http://badge.fury.io/js/bpk-stylesheets) |
 | [`bpk-svgs`](/packages/bpk-svgs) | [![npm version](https://badge.fury.io/js/bpk-svgs.svg)](http://badge.fury.io/js/bpk-svgs) |
 | [`bpk-tokens`](/packages/bpk-tokens) | [![npm version](https://badge.fury.io/js/bpk-tokens.svg)](http://badge.fury.io/js/bpk-tokens) |
+| [`react-native-bpk-component-text`](/native/packages/react-native-bpk-component-text) | [![npm version](https://badge.fury.io/js/bpk-tokens.svg)](http://badge.fury.io/js/react-native-bpk-component-text) |
```
Comment (matthewdavidson) [inline readme.md]: `https://badge.fury.io/js/bpk-tokens.svg` should be `https://badge.fury.io/js/react-native-bpk-component-text.svg` :P

Comment (shaundon) [reply readme.md]: Dammit, I knew I'd miss a link somewhere! Fixed now.
